File: idea-plugin/src/main/kotlin/dev/djarvur/c4drill/core/RenderParams.kt
// View state and c4drill/renderDiagram parameter construction for the live
// preview tool window (issue #29 scope update; mirrors the VS Code
// extension's src/core/renderParams.ts from #27). The wire contract is
// documented on internal/lsp RenderDiagramParams / RenderDiagramResult:
//
//	{ textDocument: { uri }, target, allExpanded, expanded, legend, format }
//
//	target       "" = C1 context, one segment = that unit's C2, deeper = C3
//	allExpanded  the single all-nested diagram (--expanded); overrides target
//	expanded     replacement for [properties].expanded C1 drill-down set;
//	             an EMPTY array is meaningful ("collapse all") — the field is
//	             deliberately sent even when empty, while "model default" is
//	             expressed by OMITTING the field (null here)
//	legend       true/false overrides properties.legend; null = default
//	format       only "svg" in v1
//
// Pure logic, no IntelliJ imports, so it is unit-testable with plain JUnit.

package dev.djarvur.c4drill.core

enum class LegendChoice { DEFAULT, ON, OFF }

data class PreviewViewState(
    /** Render target path; "" is the C1 context diagram. */
    val target: String = "",
    /** All-expanded diagram mode (the CLI --expanded mode); overrides target. */
    val allExpanded: Boolean = false,
    /** Comma/whitespace-separated unit paths replacing [properties].expanded; empty = model default. */
    val expandedText: String = "",
    /**
     * Explicit "collapse all" override: sends expanded: [] (an EMPTY array is
     * meaningful on the wire — the server treats it as "collapse everything").
     * Takes precedence over expandedText; cleared as soon as the user edits
     * the expanded-set input.
     */
    val collapseAll: Boolean = false,
    val legend: LegendChoice = LegendChoice.DEFAULT,
)

data class RenderDiagramWireParams(
    val textDocumentUri: String,
    val target: String? = null,
    val allExpanded: Boolean? = null,
    /** null = model default; non-null (possibly empty list) = explicit override. */
    val expanded: List<String>? = null,
    val legend: Boolean? = null,
    val format: String? = null,
)

const val renderDiagramFormat: String = "svg"

val initialViewState: PreviewViewState = PreviewViewState()

/** buildRenderParams maps view state + document URI to the renderDiagram request params. */
fun buildRenderParams(uri: String, state: PreviewViewState): RenderDiagramWireParams {
    var params = RenderDiagramWireParams(textDocumentUri = uri, format = renderDiagramFormat)

    params = if (state.allExpanded) {
        params.copy(allExpanded = true)
    } else {
        params.copy(target = state.target)
    }

    params = if (state.collapseAll) {
        // The wire contract: an empty array IS the "collapse all" override.
        params.copy(expanded = emptyList())
    } else {
        val expanded = parseExpandedSet(state.expandedText)

        if (expanded != null) params.copy(expanded = expanded) else params
    }

    return when (state.legend) {
        LegendChoice.ON -> params.copy(legend = true)
        LegendChoice.OFF -> params.copy(legend = false)
        LegendChoice.DEFAULT -> params
    }
}

/** parseExpandedSet splits the expanded-set input on commas and whitespace; blank input yields null (model default). */
fun parseExpandedSet(text: String): List<String>? {
    val entries = text.split(Regex("[\\s,]+"))
        .map { it.trim() }
        .filter { it.isNotEmpty() }

    return if (entries.isEmpty()) null else entries
}

data class BreadcrumbEntry(
    val label: String,
    val target: String,
    val current: Boolean,
)

/** breadcrumbTrail returns the clickable ancestor chain for the current target: the C1 root plus one entry per segment. */
fun breadcrumbTrail(state: PreviewViewState): List<BreadcrumbEntry> {
    val trail = mutableListOf(BreadcrumbEntry("C1", "", !state.allExpanded && state.target.isEmpty()))

    if (state.allExpanded) {
        trail.add(BreadcrumbEntry("expanded", "", current = true))

        return trail
    }

    if (state.target.isEmpty()) {
        return trail
    }

    val segments = state.target.split('.')

    for (i in segments.indices) {
        val t = segments.subList(0, i + 1).joinToString(".")

        trail.add(BreadcrumbEntry(segments[i], t, current = i == segments.size - 1))
    }

    return trail
}

File: extension/src/core/renderParams.ts
// View state and c4drill/renderDiagram parameter construction for the live
// preview panel (issue #27 scope update). The wire contract is documented on
// internal/lsp RenderDiagramParams / RenderDiagramResult:
//
//	{ textDocument: { uri }, target, allExpanded, expanded, legend, format }
//
//	target       "" = C1 context, one segment = that unit's C2, deeper = C3
//	allExpanded  the single all-nested diagram (--expanded); overrides target
//	expanded     replacement for [properties].expanded C1 drill-down set;
//	             an EMPTY array is meaningful ("collapse all") — the field is
//	             deliberately sent even when empty, while "model default" is
//	             expressed by OMITTING the field (undefined here)
//	legend       true/false overrides properties.legend; undefined = default
//	format       only "svg" in v1
//
// Pure logic, no VS Code imports, so it is unit-testable with node:test.

export type LegendChoice = 'default' | 'on' | 'off';

export interface PreviewViewState {
    /** Render target path; '' is the C1 context diagram. */
    readonly target: string;
    /** All-expanded diagram mode (the CLI --expanded mode); overrides target. */
    readonly allExpanded: boolean;
    /** Comma/whitespace-separated unit paths replacing [properties].expanded; empty = model default. */
    readonly expandedText: string;
    /**
     * Explicit "collapse all" override: sends expanded: [] (an EMPTY array is
     * meaningful on the wire — the server treats it as "collapse everything").
     * Takes precedence over expandedText; cleared as soon as the user edits
     * the expanded-set input.
     */
    readonly collapseAll: boolean;
    readonly legend: LegendChoice;
}

export const initialViewState: PreviewViewState = {
    target: '',
    allExpanded: false,
    expandedText: '',
    collapseAll: false,
    legend: 'default',
};

export interface RenderDiagramParams {
    textDocument: { uri: string };
    target?: string;
    allExpanded?: boolean;
    /** Absent = model default; present (possibly []) = explicit override. */
    expanded?: string[];
    legend?: boolean;
    format?: string;
}

export const renderDiagramFormat = 'svg';

// buildRenderParams maps view state + document URI to the renderDiagram
// request params, honoring the empty-array-means-collapse-all contract.
export function buildRenderParams(uri: string, state: PreviewViewState): RenderDiagramParams {
    const params: RenderDiagramParams = {
        textDocument: { uri },
        format: renderDiagramFormat,
    };

    if (state.allExpanded) {
        params.allExpanded = true;
    } else {
        params.target = state.target;
    }

    if (state.collapseAll) {
        // The wire contract: an empty array IS the "collapse all" override.
        params.expanded = [];
    } else {
        const expanded = parseExpandedSet(state.expandedText);
        if (expanded !== undefined) {
            params.expanded = expanded;
        }
    }

    if (state.legend === 'on') {
        params.legend = true;
    } else if (state.legend === 'off') {
        params.legend = false;
    }

    return params;
}

// parseExpandedSet splits the expanded-set input on commas and whitespace.
// Empty/blank input yields undefined (model default); non-empty input always
// yields an array (possibly empty entries are filtered — explicit entries
// only). An input of only separators (",,") therefore resets to the model
// default, while listing zero units is not expressible via the text box
// (the UI's "Collapse all" button sends the empty-array override instead).
export function parseExpandedSet(text: string): string[] | undefined {
    const entries = text
        .split(/[\s,]+/)
        .map((s) => s.trim())
        .filter((s) => s !== '');

    return entries.length === 0 ? undefined : entries;
}

// breadcrumbTrail returns the clickable ancestor chain for the current
// target: the C1 root entry plus one entry per dotted segment. The last
// entry is the current view (not navigable).
export interface BreadcrumbEntry {
    readonly label: string;
    readonly target: string;
    readonly current: boolean;
}

export function breadcrumbTrail(state: PreviewViewState): BreadcrumbEntry[] {
    const trail: BreadcrumbEntry[] = [
        { label: 'C1', target: '', current: !state.allExpanded && state.target === '' },
    ];

    if (state.allExpanded) {
        trail.push({ label: 'expanded', target: '', current: true });

        return trail;
    }

    if (state.target === '') {
        return trail;
    }

    const segments = state.target.split('.');

    for (let i = 0; i < segments.length; i++) {
        const t = segments.slice(0, i + 1).join('.');

        trail.push({ label: segments[i], target: t, current: i === segments.length - 1 });
    }

    return trail;
}
